### PR TITLE
Added support for querying objects by metadata fields

### PR DESCRIPTION
**Pull request by LeOndaz**
I'm leaving this PR here as in case you wanted to discuss the approach with me, I'll be adding tests and generalizing the metadata for all models that need it soon and I'll also rebuild the schema as the tests below failed because of it, just lemme hear your thoughts

### CODE REVIEW

**Comment by LeOndaz** [comment]
I'd appreciate suggestions about this test should be implemented

File: saleor/graphql/core/filters.py
Comment (maarcingebala) [inline saleor/graphql/core/filters.py:81]: This is correct, but I would suggest using an f-string for better readability:
```suggestion
        lookup = f'metadata__{metadata_item.key}'
```

**Review by maarcingebala — COMMENTED** [review]
Tests for metadata API can be found in `saleor/graphql/meta/tests`. We could add them to `test_meta_queries.py`, but this file is pretty large and should be split into smaller files, so these tests could also be added to a new file e.g. `test_meta_filtering.py`. In each test we could take an object from a queryset, add a metadata key and value; then run a GraphQL query that would filter out this object.

File: saleor/graphql/core/filters.py
Comment (LeOndaz) [reply saleor/graphql/core/filters.py:81]: @maarcingebala I've used fstrings at the beginning but I said to myself I don't know what's the least python version you're supporting, so I chose the one that's more supported, I'll change it for sure!

Comment (LeOndaz) [reply saleor/graphql/core/filters.py:81]: And thanks for the hints, I'll do them, can you have a look at #6684

**Comment by LeOndaz** [comment]
The test is added, it should pass after the checks are finished. Please don't forget to have a look at #6684 as this is important for my next PRs.

File: saleor/graphql/product/filters.py
Comment (korycins) [inline saleor/graphql/product/filters.py:321]: I would suggest creating the MetadataFilterBase class, inherited from django_filters.FilterSet. It should be placed somewhere in the `core` directory.  Then all filtres which should have metada field can use this class to automatically include it. Something similar we have for db models where we have `ModelWithMetadata` class to include meta fields. 
It should be similar to this example:
```
class Test(django_filters.FilterSet):
    metadata = MetadataFilter()
    class Meta:
        abstract=True

class ProductFilter(Test):
```

Comment (LeOndaz) [reply saleor/graphql/product/filters.py:321]: At first I did that, but I didn't do it this way, probably because I'm still convinced that Metadata API is somewhat weird, to get my point can you have a look at the mentioned issue?

**Comment by LeOndaz** [comment]
Any ideas on the adding more tests ?

**Comment by LeOndaz** [comment]
I have an idea in mind that I'd like to share, I'm thinking of extending this search to a better version
```
metadata: {
   items: [],
   searchAlgorithm: CONCAT_RESULTS, FILTER_STACK, WHATEVER,
   keyCase: MATCH_CASE, ANY_CASE,
   value: ... 
}
```
and if value is provided, we assume it's the value, if not, we assume it has no value and it's an empty metadata item.
Just an overview of what I'm thinking, what do you think? ignoring the weird names, will rethink them if we agree on them

**Comment by LeOndaz** [comment]
Also before putting effort in this, I'd like you to have a look at the mentioned issue kindly to see what I'm thinking

**Comment by korycins** [comment]
The one thing that I am worried about is the performance of DB, when Saleor will have a lot of products. 
@patrys  what do you think?

**Comment by patrys** [comment]
I agree. We don't want to filter on non-indexed data. We've previously kept metadata in a JSONB structure, but is there anything stopping us from using an indexed HSTORE?

**Comment by LeOndaz** [comment]
@patrys Tbh I've never used HSTOREs, what do you think is the possible future of this PR ? if it may be accepted then we may in the future change the implementation itself while providing the same interface (or whatever), I'm stuck in my project because there's no way to add featured categories, collections, ...etc this is actually why I PRed this

**Comment by patrys** [comment]
I understand how this implementation is already helpful in your case, but I'm hesitant to merge anything that we already know won't scale for at least some of our users. This is a great feature to have, but I believe we should tackle it in a performance-friendly way (which could mean migrating from JSONB to an indexed HSTORE column).

**Comment by LeOndaz** [comment]
@patrys totally it's totally fine, if I were you I'd just do the same so I understand, I may help with this soon after I finish my current project s & PRs. Have a nice day!

**Comment by patrys** [comment]
@LeOndaz see #6979 as it's adding indexes that make `metadata__contains=...` lookups fast with the existing JSONB fields.

File: saleor/graphql/core/filters.py
Comment (patrys) [inline saleor/graphql/core/filters.py:76]:
```suggestion
        qs = qs.filter(metadata__contains={metadata_item.key: metadata_item.value})
```

**Comment by LeOndaz** [comment]
Great! here we go, I'll do it maybe later today

**Review by patrys — COMMENTED** [review]
I would love to make sure we get the expected SQL execution plan (e.g., EXPLAIN not including any seq scans) for queries generated by this filter (especially when combined with another filter).

**Comment by LeOndaz** [comment]
We may have a nice test for this, unfortunately, I'm having my exams these days, after I finish I may do it, you can merge this (if you see it's okay) and the test will be in a separate PR

**Review by korycins — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Comment by maarcingebala** [comment]
@LeOndaz Would you have some capacity to fix the linters? I should only require running `Black` and submitting reformatted code. After that, I think this PR should be ready to merge.

**Comment by LeOndaz** [comment]
@maarcingebala sure, it's done

**Comment by korycins** [comment]
@LeOndaz  It seems that you pushed the incorrect branch here

**Comment by LeOndaz** [comment]
Sorry about that, I didn't work on this for weeks and I forgot where goes where, anyway I think I've done it right now

**Comment by LeOndaz** [comment]
I thought I've wasted the files actually, I did reflog the repo to see if it's still there and I got it from there, you'll need to have a look

**Comment by LeOndaz** [comment]
I think some other commits I've done before was pushed with this, I'll mention you when everything is resolved

**Comment by LeOndaz** [comment]
@maarcingebala I've done it, everything should be working yet I've absolutely zero idea about what's failing, I've read it and saw some code like `assert 4 == 5` which is weird, anyway, I've updated saleor version to the master one too, so this is coming directly from master branch

**Review by fowczarek — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Review by d-wysocki — APPROVED** [review]

**Review by tomaszszymanski129 — APPROVED** [review]

**Comment by maarcingebala** [comment]
@LeOndaz Thank you for contributing to Saleor 🎉

**Comment by LeOndaz** [comment]
I hope for even more than contributions someday, have a great day!

**Pull request merged**
